### PR TITLE
add encoding module

### PR DESCRIPTION
**Pull request by fubhy**
Re-opening this here.

#### Benchmark  (Custom [this] vs. Node.js `Buffer`) [on my laptop]

Encoding: This implementation and the Node.js `Buffer` API perform equally
Decoding: The Node.js `Buffer` API performs ~30% better for decoding large strings (I previously tested it with short strings where the difference was insigificant).

However, please note that the Buffer API behaves recklessly when passed invalid hex / base64 strings. Instead of erroring out, it simply ignores / skips invalid characters.

I'll personally happily sacrifice 30% performance for correctness.

```
  Custom - test/Encoding.bench.ts > Hex > Encode
    1.07x faster than Buffer

  Buffer - test/Encoding.bench.ts > Hex > Decode
    1.39x faster than Custom

  Custom - test/Encoding.bench.ts > Base64 > Encode
    1.00x faster than Buffer

  Buffer - test/Encoding.bench.ts > Base64 > Decode
    1.31x faster than Custom
```

### CODE REVIEW

File: src/internal/Encoding/Base64.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+  }
+
+  return result
+}
+
+/** @internal */
+export const decode = (str: string) => {
+  const length = str.length
+  if (length % 4 !== 0) {
+    throw new TypeError("Invalid base64 string")
```
Comment (fubhy) [inline src/internal/Encoding/Base64.ts:37]: @tim-smart If we return `Either` here, should I also remove the (then somewhat superfluous) `unsafeDecode` variants?

Comment (tim-smart) [reply src/internal/Encoding/Base64.ts]: Yes that keeps the module simple.

File: src/Encoding.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * @category encoding
+ * @since 1.0.0
+ */
+export const encodeBase64Url: (bytes: Uint8Array) => string = Base64Url.encode
+
+/**
+ * Decodes a base64 (URL) encoded string.
+ *
+ * @category encoding
```
Comment (tim-smart) [inline src/Encoding.ts:45]: Using a category of "decoding" for the decode apis would be a nice touch.

File: src/internal/Encoding/Base64.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    result += base64abc[(bytes[i - 1] & 0x0f) << 2]
+    result += "="
+  }
+
+  return result
+}
+
+/** @internal */
+export class Base64DecodeError {
+  readonly _tag = "Base64DecodeError"
```
Comment (tim-smart) [inline src/internal/Encoding/Base64.ts:37]: I think adding a bit more context to the errors would be helpful. Maybe a copy of the offending input and a message?

**Review by tim-smart — COMMENTED** [review]
Looking pretty good!

I'm not too worried about the performance - it is still really easy to use the Buffer api on node.js if you don't want to use this module.

File: src/internal/Encoding/Base64.ts (hunk 3 above)
Comment (fubhy) [reply src/internal/Encoding/Base64.ts:37]: @tim-smart Done! Let me know if you think the messages are descriptive enough.

Comment (tim-smart) [reply src/internal/Encoding/Base64.ts:37]: Yes that is better :)

I think we should just have one error type for the entire module, and it should also be exported alongside a constructor. It should also have it's own TypeId in the same fashion as other error in the ecosystem (for example https://effect-ts.github.io/io/modules/Cause.ts.html#nosuchelementexception-interface). This prevents typescript from merging it, in the small chance someone decides to use the same _tag in another error.

Once errors are sorted I think we can merge this!

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: src/internal/Encoding/Base64.ts (hunk 3 above)
Comment (fubhy) [reply src/internal/Encoding/Base64.ts]: Done!

**Comment by fubhy** [comment]
@tim-smart I kept the name `DecodeException`. Should we maybe call this "IncorrectEncodingException` or `InvalidEncodingException`?

File: src/internal/Encoding/Base64Url.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  if (!/^[-_A-Z0-9]*?={0,2}$/i.test(str)) {
+    return Either.left(DecodeException(str, "Invalid input"))
+  }
+
+  // Some variants allow or require omitting the padding '=' signs
+  let sanitized = length % 4 === 2 ? `${str}==` : length % 4 === 3 ? `${str}=` : str
+  sanitized = sanitized.replace(/-/g, "+").replace(/_/g, "/")
+
+  return Base64.decode(sanitized).pipe(Either.mapLeft((e) => DecodeException(e.input, e.message)))
```
Comment (tim-smart) [inline src/internal/Encoding/Base64Url.ts:27]: You don't need this mapLeft anymore.

**Review by tim-smart — CHANGES_REQUESTED** [review]

**Review by tim-smart — APPROVED** [review]

**Pull request merged**